### PR TITLE
Recurring salesorder

### PR DESCRIPTION
**Pull request by interhawk**
Client code and ui forms to enable creating recurring sales orders.

### CODE REVIEW

File: guiclient/salesOrder.cpp [hunk 1]
```diff
@@ -729,6 +735,43 @@ void salesOrder::sSave()
...
+      {
+        // Update Recurring Sales Order with latest saved detail
+        XSqlQuery recurUpd;
+        recurUpd.prepare("UPDATE recur SET recur_parent_id=:newso "
+          "WHERE ((recur_parent_type = 'SALESORDER') "
+          " AND   (recur_parent_id = :oldso));");
+        recurUpd.bindValue(":newso", _soheadid);
+        recurUpd.bindValue(":oldso", _recur->parentId());
+        recurUpd.exec();
+        if (recurUpd.lastError().type() != QSqlError::NoError)
```
Comment (andersdd) [inline guiclient/salesOrder.cpp]: ErrorReporter checks for sql errors.  You should be able to just go:
```
if(ErrorReporter::error(QtCriticalMsg, this, tr("Error Updating Sales Order Recurrence"),
            recurUpd, __FILE__, __LINE__))
          return;
```

File: guiclient/salesOrder.cpp [hunk 2]
```diff
@@ -729,6 +735,43 @@ void salesOrder::sSave()
...
+          ErrorReporter::error(QtCriticalMsg, this, tr("Error Updating Sales Order Recurrence"),
+            recurUpd, __FILE__, __LINE__);
+          return;
+        }
+        // Also update this Sales Order
+        recurUpd.prepare("UPDATE cohead SET cohead_recurring_cohead_id=:salesorder "
+          "WHERE (cohead_id = :salesorder); ");
+        recurUpd.bindValue(":salesorder", _soheadid);
+        recurUpd.exec();
+        if (recurUpd.lastError().type() != QSqlError::NoError)
```
Comment (andersdd) [inline guiclient/salesOrder.cpp]: Simplify error handling

File: guiclient/salesOrder.cpp [hunk 3]
```diff
@@ -774,6 +817,43 @@ void salesOrder::sSaveAndAdd()
...
+      {
+        // Update Recurring Sales Order with latest saved detail
+        XSqlQuery recurUpd;
+        recurUpd.prepare("UPDATE recur SET recur_parent_id=:newso "
+          "WHERE ((recur_parent_type = 'SALESORDER') "
+          " AND   (recur_parent_id = :oldso));");
+        recurUpd.bindValue(":newso", _soheadid);
+        recurUpd.bindValue(":oldso", _recur->parentId());
+        recurUpd.exec();
+        if (recurUpd.lastError().type() != QSqlError::NoError)
```
Comment (andersdd) [inline guiclient/salesOrder.cpp]: Simplify error handling

File: guiclient/salesOrder.cpp [hunk 4]
```diff
@@ -2742,6 +2848,10 @@ void salesOrder::populate()
           _fromQuote->setText(so.value("rahead_number").toString());
         }
       }
+      if (so.value("cohead_recurring_cohead_id").toInt() != 0)
+        _recur->setParent(so.value("cohead_recurring_cohead_id").toInt(), "SALESORDER");
+      else
+        _recur->setParent(_soheadid, "SALESORDER");
```
Comment (andersdd) [inline guiclient/salesOrder.cpp]: Sales Order code "S" on these lines

File: guiclient/salesOrder.ui [hunk 5]
```diff
@@ -865,7 +874,16 @@ to be bound by its terms.</comment>
...
+             <property name="leftMargin">
+              <number>0</number>
+             </property>
+             <property name="topMargin">
+              <number>0</number>
+             </property>
+             <property name="rightMargin">
+              <number>0</number>
+             </property>
+             <property name="bottomMargin">
```
Comment (andersdd) [inline guiclient/salesOrder.ui]: The editor defines margins quite verbosely and tends to override existing settings.  You can simplify margins to 
```
             <property name="margin">
              <number>0</number>
             </property>
```

File: guiclient/createRecurringItems.cpp [hunk 6]
```diff
@@ -64,11 +67,12 @@ void createRecurringItems::sCreate()
...
-    { _vouchers,   "V"     },
-    { _incidents,  "INCDT" },
-    { _projects,   "J"     },
-    { _todoItems,  "TODO"  }
+    { _invoices,   "I"         },
+    { _vouchers,   "V"         },
+    { _incidents,  "INCDT"     },
+    { _projects,   "J"         },
+    { _todoItems,  "TODO"      },
+    { _salesOrder, "SALESORDER"}
```
[the remark this answers is not in the document]
Comment (interhawk) [reply guiclient/createRecurringItems.cpp]: implemented

File: guiclient/createRecurringItems.cpp [hunk 7]
```diff
@@ -117,11 +121,12 @@ void createRecurringItems::sSubmit()
...
-    { _vouchers,   "V"     },
-    { _incidents,  "INCDT" },
-    { _projects,   "J"     },
-    { _todoItems,  "TODO"  }
+    { _invoices,   "I"          },
+    { _vouchers,   "V"          },
+    { _incidents,  "INCDT"      },
+    { _projects,   "J"          },
+    { _todoItems,  "TODO"       },
+    { _salesOrder, "SALESORDER" }
```
[the remark this answers is not in the document]
Comment (interhawk) [reply guiclient/createRecurringItems.cpp]: implemented

File: guiclient/createRecurringItems.cpp [hunk 8]
```diff
@@ -143,6 +148,8 @@ void createRecurringItems::sSubmit()
         params.append("action_name", "CreateRecurringProjects");
       else if (list[i].arg == "TODO")
         params.append("action_name", "CreateRecurringTodos");
+      else if (list[i].arg == "SALESORDER")
```
[the remark this answers is not in the document]
Comment (interhawk) [reply guiclient/createRecurringItems.cpp]: implemented

File: guiclient/salesOrder.cpp [hunk 9]
```diff
@@ -514,6 +519,7 @@ enum SetResponse salesOrder::set(const ParameterList &pParams)
       populateCMInfo();
       populateCCInfo();
       sFillCcardList();
+      _recur->setParent(_soheadid, "SALESORDER");
```
[the remark this answers is not in the document]
Comment (interhawk) [reply guiclient/salesOrder.cpp]: implemented

File: guiclient/salesOrder.cpp (hunk 1 above)
Comment (interhawk) [reply guiclient/salesOrder.cpp]: implemented

File: guiclient/salesOrder.cpp (hunk 2 above)
Comment (interhawk) [reply guiclient/salesOrder.cpp]: implemented

File: guiclient/salesOrder.cpp (hunk 3 above)
Comment (interhawk) [reply guiclient/salesOrder.cpp]: implemented

File: guiclient/salesOrder.cpp (hunk 4 above)
Comment (interhawk) [reply guiclient/salesOrder.cpp]: implemented

File: guiclient/salesOrder.ui (hunk 5 above)
Comment (interhawk) [reply guiclient/salesOrder.ui]: implemented

File: guiclient/salesOrder.cpp [hunk 10]
```diff
@@ -1266,10 +1351,15 @@ bool salesOrder::save(bool partial)
...
+
   saveSales.exec();
-  if (ErrorReporter::error(QtCriticalMsg, this, tr("Saving Order"),
-                           saveSales, __FILE__, __LINE__))
+  if (saveSales.lastError().type() != QSqlError::NoError)
   {
+    rollbackq.exec();
+    ErrorReporter::error(QtCriticalMsg, this, tr("Error Saving Order"),
+      saveSales, __FILE__, __LINE__);
     return false;
```
Comment (andersdd) [inline guiclient/salesOrder.cpp]: The above does not look correct.  The ErrorReporter should still be the logic in the `if` statement

File: guiclient/salesOrder.ui [hunk 11]
```diff
@@ -360,7 +360,7 @@ to be bound by its terms.</comment>
    <item>
     <widget class="QTabWidget" name="_salesOrderInformation">
      <property name="currentIndex">
-      <number>0</number>
+      <number>9</number>
      </property>
```
Comment (andersdd) [inline guiclient/salesOrder.ui]: This does not look correct.  The editor (unfortunately) saves the currently active Tab as the currentIndex when editing a UI.  Make sure you set currentIndex to 0 (or save the UI with the first tab active).

File: guiclient/salesOrder.cpp (hunk 10 above)
Comment (gilmoskowitz) [reply guiclient/salesOrder.cpp]: This looks correct as it stands. The `if (sql error occurred) then rollback & error report` prevents holding the transaction open while waiting for user acknowledgement of the failure.

**Comment by andersdd** [comment]
OK to merge when xtuple/xtuple#3461 ready

**Pull request merged**